spec/components/cards_spec.rb

```diff
@@ -0,0 +1,280 @@
+require "rails_helper"
+
+describe "Cards", type: :view do
+  def component_name
+    "cards"
+  end
+
+  it "renders nothing when no data is provided" do
+    assert_empty render_component({})
+  end
+
+  it "renders component wrapper" do
+    test_data = {
+      items: [
+        {
+          link: {
+            text: "Benefits",
+            path: "http://www.gov.uk",
+          },
+        },
+      ],
+    }
+    render_component(test_data)
+    assert_select ".gem-c-cards", count: 1
+  end
+
+  it "renders list heading" do
+    test_data = {
+      heading: "Topics",
+      items: [
+        {
+          link: {
+            text: "Benefits",
+            path: "http://www.gov.uk",
+          },
+        },
+      ],
+    }
+    render_component(test_data)
+    assert_select ".gem-c-cards__heading.govuk-heading-m", text: "Topics"
+  end
+
+  it "renders list heading using default heading level" do
```

Comment:
Ideally unit tests should test a single logical outcome. So if the default heading level functionality breaks, having two isolated tests will assert that it is the broken default heading level that needs fixing whilst the heading classes will still be asserted to be correct. 

And if the requirement to use a default heading level here was to be removed in the future and this test was called `renders list heading using default heading level`, it would be easy for the developer to think that the test can be removed and overlook the fact that it is also used to assert the correct heading classes.

Reply:
I can see how the level of test granularity if very different for this repo compared to `govuk-frontend` – maybe a point of discussion for one of the following community meetings, as we currently don't have any specific guidance on what good looks like for an rspec test (at least none that I'm aware of).

Reply:
Following our community discussion, we agree that there is a currently undocumented convention with GOV.UK Publishing components to only test markup where the template contains logic (eg. conditionals or forks), rather than testing for the integrity of the generated markup eg. that the correct classes and elements are generated even where there is no explicit template logic involved.

We do have visual regression tests and do PR reviews so those give us additional confidence.

However we might wish to consider testing for the integrity of the markup in the future, especially if we saw instances where this could have flagged unwanted changes. I've removed some tests to align with our current convention but left their removal out as a separate commit in case we want to revisit this later.